indexdata.go

```diff
@@ -387,32 +407,30 @@ func (d *indexData) iterateNgrams(query *query.Substring) (*ngramIterationResult
 	slices.SortFunc(ngramOffs, func(a, b runeNgramOff) bool {
 		return a.ngram < b.ngram
 	})
-	frequencies := make([]uint32, 0, len(ngramOffs))
-	ngramLookups := 0
-	for _, o := range ngramOffs {
-		var freq uint32
-		if query.CaseSensitive {
-			freq = d.ngramFrequency(o.ngram, query.FileName)
-			ngramLookups++
-		} else {
-			for _, v := range generateCaseNgrams(o.ngram) {
-				freq += d.ngramFrequency(v, query.FileName)
-				ngramLookups++
-			}
-		}
 
-		if freq == 0 {
-			return &ngramIterationResults{
-				matchIterator: &noMatchTree{
-					Why: "freq=0",
-					Stats: Stats{
-						NgramLookups: ngramLookups,
-					},
+	ngrams := make([]ngram, 0, len(ngramOffs))
+	for _, ng := range ngramOffs {
+		ngrams = append(ngrams, ng.ngram)
+	}
+
+	// Look up ngram indexes without loading posting lists. This way we can stop
+	// early if a ngram does not exist. On the flip side we incur an additional
+	// loop.
+	ngramIndexes, ngramLookups := d.ngramIndexes(ngrams, query.FileName, query.CaseSensitive)
+	if len(ngramIndexes) == 0 {
+		return &ngramIterationResults{
+			matchIterator: &noMatchTree{
+				Why: "freq=0",
+				Stats: Stats{
+					NgramLookups: ngramLookups,
 				},
-			}, nil
-		}
+			},
+		}, nil
+	}
 
-		frequencies = append(frequencies, freq)
+	frequencies := make([]uint32, 0, len(ngramOffs))
+	for _, ngramIndex := range ngramIndexes {
+		frequencies = append(frequencies, d.ngramIndexFrequency(ngramIndex, query.FileName))
```

Comment:
you might as well make ngramIndexFrequency take in the ngramIndexes slice and avoid the loop here.